include/aspect/fe_variable_system.h

```diff
@@ -0,0 +1,423 @@
+   * information like a name and the block structure used in linear systems. A
+   * vector of these structs will be converted into an ordered collection of
+   * objects of type an FEVariable represented by FEVariableSystem.
+   *
+   * Blocks are used to extract or solve for certain variables separately or
+   * in block based preconditioners.  As an example, in a Stokes problem one
+   * might want velocity and pressure in a single block (for use with a direct
+   * solver), or one block for all velocity components and a second block for
+   * the pressure (for a Schur complement-based preconditioner), or separate
+   * blocks for pressure and each velocity component.
+   */
+  template <int dim>
+  struct VariableDefinition
+  {
+    /**
+     * Constructor.
+     *
+     * @param name A user-friendly and unique string representation of the variable.
+     *
+     * @param fe The FiniteElement class to use.
+     *
+     * @param multiplicity Number of copies of the @p fe to create.
+     *
+     * @param n_blocks Number of blocks this variable represents inside the
+     * linear system. A value of 0 will add the next variable to the current
+     * block, while 1 will put the next variable into a new block. A value
+     * that is equal to the number of components will create a block for each
+     * component.
+     */
+    VariableDefinition(const std::string &name,
+                       std_cxx11::shared_ptr<FiniteElement<dim> > fe,
+                       const unsigned int multiplicity,
+                       const unsigned int n_blocks);
+
+    /**
+     * Return the total number of components of this variable.
+     */
+    unsigned int n_components() const;
+
+    /**
+     * Name of the variable used in FEVariableSystem to identify it.
+     */
+    std::string name;
+    /**
+     * The FiniteElement space.
+     */
+    std_cxx11::shared_ptr<FiniteElement<dim> > fe;
+    /**
+     * The multiplicity used in FESystem: how many copies of @p fe are there?
+     */
+    unsigned int multiplicity;
+    /**
+     * Number of linear algebra blocks occupied by this variable. See
+     * constructor for details.
+     */
+    unsigned int n_blocks;
+  };
+
+  /**
+   * Struct to represent a variable as part of a FEVariableSystem. Constructed
+   * from an instance of VariableDefinition<dim> but contains additional
+   * information that can be queried.
+   */
+  template<int dim>
+  struct FEVariable: public VariableDefinition<dim>
+  {
+      /**
+       * Initialize this variable as part of a FE_System with the given
+       * indices for @p component, @p block, and @p base.
+       */
+      FEVariable(const VariableDefinition<dim> &fe_variable,
+                 const unsigned int component_index,
+                 const unsigned int block_index,
+                 const unsigned int base_index);
+
+      unsigned int first_component_index;
+      unsigned int block_index;
+      unsigned int base_index;
```

Comment:
Yes, but then I don't have the default ``operator=()``, which is need to use ``reserve()`` in the ``std::vector``.  Or am I missing something?